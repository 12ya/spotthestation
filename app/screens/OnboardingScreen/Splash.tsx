import React, { useEffect } from "react"
import { ImageBackground, ViewStyle, Image, ImageStyle, View, TextStyle } from "react-native"
import { useNavigation } from "@react-navigation/native"
import { Screen, Text } from "../../components"
import { colors } from "../../theme/colors"
import { typography } from "../../theme/typography"
import { useSafeAreaInsetsStyle } from "../../utils/useSafeAreaInsetsStyle"
import { IconLinkButton } from "./components/IconLinkButton"
import { NasaLogo } from "./components/NasaLogo"
import * as storage from "../../utils/storage"
import { getUserId } from "../../utils/generateUUID"
import { fontSizes, lineHeights, scale } from "../../theme/spacing"

const background = require("../../../assets/images/bg.png")
const iss = require("../../../assets/images/iss-with-path.png")

export function Splash() {
  const navigation = useNavigation()
  
  const $topInset = useSafeAreaInsetsStyle(["top", "bottom"], "padding")

  const generateUser = async () => {
    if ((!await storage.load("userId"))) {
      await storage.save("userId", getUserId())
      setTimeout(() => {
        handleNavigate()
      }, 3000)
    }
  }

  useEffect(() => {
    generateUser().catch(e => console.log(e))
  }, [])

  const handleNavigate = () => navigation.navigate("SignupCompleteProfile" as never)

  return (
    <Screen preset="fixed" contentContainerStyle={$container} statusBarStyle="light">
      <ImageBackground source={background} resizeMode="cover" style={[$imageBackground, $topInset]}>
        <NasaLogo />
        <Image source={iss} style={$iss} />
        <View style={$bottomContainer}>
          <Text
            accessible
            accessibilityLabel="Splash title"
            accessibilityHint="Splash screen title"
            accessibilityRole="text"
            tx="onboarding.splash.title"
            style={$title}
          />
          <View style={$bottomRowContainer}>
            <Text
              accessible
              accessibilityLabel="Splash subtitle"
              accessibilityHint="Splash screen subtitle"
              accessibilityRole="text"
              tx="onboarding.splash.subTitle"
              style={$subTitle}
            />
            <IconLinkButton 
              icon="back"
              accessible
              accessibilityLabel="Arrow right button"
              accessibilityHint="Navigates to the next screen"
              buttonStyle={$buttonSize} 
              imageStyle={$imageStyle} 
              onPress={handleNavigate}
            />
          </View>
        </View>
      </ImageBackground>
    </Screen>
  )
}

const $container: ViewStyle = {
  flex: 1,
}

const $imageBackground: ViewStyle = {
  flex: 1,
  display: "flex",
  flexDirection: "column"
}

const $bottomContainer: ViewStyle = {
  position: "absolute",
  top: "68%"
}

const $bottomRowContainer: ViewStyle = {
  flexDirection: "row",
  alignItems: "center",
  justifyContent: "space-between",
  width: "100%",
  paddingHorizontal: scale(36)
}

const $iss: ImageStyle = {
  position: "absolute",
  top: "26%"
}

const $title: TextStyle = {
  width: scale(220),
  marginLeft: scale(36),
  fontSize: fontSizes[48],
  lineHeight: lineHeights[64],
  fontFamily: typography.primary.bold,
  color: colors.palette.neutral100,
}

const $subTitle: TextStyle = {
  width: '85%',
  fontSize: fontSizes[24],
  lineHeight: lineHeights[29],
  fontFamily: typography.primary.normal,
  color: colors.palette.neutral250
}

const $buttonSize: TextStyle = {
  width: scale(64),
  height: scale(64)
}

const $imageStyle: ImageStyle = {
  transform: [{ rotate: "180deg" }]
}
