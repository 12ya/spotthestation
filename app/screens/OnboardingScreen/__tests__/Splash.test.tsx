import { useNavigation } from "@react-navigation/native"
import React from "react"
import { fireEvent, render } from "@testing-library/react-native"
import { Splash } from "../Splash"

jest.mock("@react-navigation/native", () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}))

jest.mock("../../../utils/storage", () => ({
  load: jest.fn(() => Promise.resolve(null)),
  save: jest.fn(() => Promise.resolve()),
}))
jest.mock("../components/IconLinkButton", () => ({
  IconLinkButton: ({ icon, onPress }) => <button onClick={onPress}>{icon}</button>,
}))

describe("Splash", () => {
  it("renders correctly", () => {
    const tree = render(<Splash />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("navigates to the next screen when Arrow right button is clicked", () => {
    const component = render(<Splash />)
    const openSettingsButton = component.root.findByProps({
      icon: "back",
    })

    fireEvent.press(openSettingsButton)
    expect(useNavigation().navigate).not.toHaveBeenCalled()
  })
})
