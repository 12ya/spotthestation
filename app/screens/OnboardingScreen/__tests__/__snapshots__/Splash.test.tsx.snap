// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Array [
        Object {
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
        },
        Object {},
      ]
    }
  >
    <Image
      resizeMode="cover"
      source={1}
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": undefined,
            "width": undefined,
          },
          undefined,
        ]
      }
    />
    <RCTImageView
      internal_analyticTag={null}
      resizeMode="cover"
      source={
        Array [
          Object {
            "height": undefined,
            "uri": undefined,
            "width": undefined,
          },
        ]
      }
      style={
        Object {
          "height": 110,
          "marginLeft": 36,
          "marginTop": 36,
          "overflow": "hidden",
          "width": 110,
        }
      }
    />
    <RCTImageView
      internal_analyticTag={null}
      resizeMode="cover"
      source={
        Array [
          Object {
            "height": undefined,
            "uri": undefined,
            "width": undefined,
          },
        ]
      }
      style={
        Object {
          "height": undefined,
          "overflow": "hidden",
          "position": "absolute",
          "top": "26%",
          "width": undefined,
        }
      }
    />
    <View
      style={
        Object {
          "position": "absolute",
          "top": "68%",
        }
      }
    >
      <div />
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingHorizontal": 36,
            "width": "100%",
          }
        }
      >
        <div />
        <View
          accessibilityHint="Navigates to the next screen"
          accessibilityLabel="Arrow right button"
          accessibilityRole="imagebutton"
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#1E2022",
                "borderRadius": 100,
                "display": "flex",
                "height": 42,
                "justifyContent": "center",
                "overflow": "hidden",
                "position": "relative",
                "width": 42,
              },
              Object {
                "height": 64,
                "width": 64,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                },
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#1E2022",
                    "borderRadius": 100,
                    "display": "flex",
                    "height": 42,
                    "justifyContent": "center",
                    "overflow": "hidden",
                    "position": "relative",
                    "width": 42,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            <ViewManagerAdapter_ExpoBlurView
              proxiedProperties={
                Object {
                  "intensity": 0,
                  "tint": "default",
                }
              }
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            />
            <div />
          </View>
        </View>
      </View>
    </View>
  </View>
</div>
`;
